If numpy is not installed, test suite fails in an unexpected way

@dean0x7d 

Python 3.5.2 via conda, gcc 5.2.1, pytest 2.9.2, Linux. This only happens if I remove numpy package from the conda environment.

```sh
collected 40 items / 9 errors

test_callbacks.py FFFF
test_constants_and_functions.py FFF
test_enum.py FFF
test_eval.py F
test_exceptions.py F
test_inheritance.py FF
test_issues.py FFFFFFFF
test_keep_alive.py FFF
test_modules.py FFF
test_numpy_vectorize.py sss
test_opaque_types.py FF
test_operator_overloading.py F
test_sequences_and_iterators.py FFF
test_stl_binders.py FFF
================================================================== short test summary info ==================================================================
SKIP [1] tests/test_numpy_vectorize.py:69: numpy is not installed
SKIP [1] tests/test_numpy_vectorize.py:60: numpy is not installed
SKIP [1] tests/test_numpy_vectorize.py:6: numpy is not installed

========================================================================== ERRORS ===========================================================================
__________________________________________________________ ERROR collecting tests/test_buffers.py ___________________________________________________________
test_buffers.py:2: in <module>
    from pybind11_tests import Matrix, ConstructorStats
E   ImportError: generic_type: type "ConstructorStats" is already registered!
___________________________________________________________ ERROR collecting tests/test_eigen.py ____________________________________________________________
test_eigen.py:7: in <module>
    ref = np.array([[ 0,  3,  0,  0,  0, 11],
E   NameError: name 'np' is not defined
____________________________________________________ ERROR collecting tests/test_kwargs_and_defaults.py _____________________________________________________
test_kwargs_and_defaults.py:2: in <module>
    from pybind11_tests import (kw_func0, kw_func1, kw_func2, kw_func3, kw_func4, call_kw_func,
E   ImportError: generic_type: type "ConstructorStats" is already registered!
___________________________________________________ ERROR collecting tests/test_methods_and_attributes.py ___________________________________________________
test_methods_and_attributes.py:1: in <module>
    from pybind11_tests import ExampleMandA, ConstructorStats
E   ImportError: generic_type: type "ConstructorStats" is already registered!
________________________________________________________ ERROR collecting tests/test_numpy_dtypes.py ________________________________________________________
test_numpy_dtypes.py:9: in <module>
    simple_dtype = np.dtype({'names': ['x', 'y', 'z'],
E   NameError: name 'np' is not defined
__________________________________________________________ ERROR collecting tests/test_pickling.py __________________________________________________________
test_pickling.py:6: in <module>
    from pybind11_tests import Pickleable
E   ImportError: generic_type: type "ConstructorStats" is already registered!
________________________________________________________ ERROR collecting tests/test_python_types.py ________________________________________________________
test_python_types.py:3: in <module>
    from pybind11_tests import ExamplePythonTypes, ConstructorStats
E   ImportError: generic_type: type "ConstructorStats" is already registered!
_________________________________________________________ ERROR collecting tests/test_smart_ptr.py __________________________________________________________
test_smart_ptr.py:1: in <module>
    from pybind11_tests import ConstructorStats
E   ImportError: generic_type: type "ConstructorStats" is already registered!
_____________________________________________________ ERROR collecting tests/test_virtual_functions.py ______________________________________________________
test_virtual_functions.py:2: in <module>
    import pybind11_tests
E   ImportError: generic_type: type "ConstructorStats" is already registered!
========================================================================= FAILURES ==========================================================================
______________________________________________________________________ test_callbacks _______________________________________________________________________

    def test_callbacks():
        from functools import partial
>       from pybind11_tests import (test_callback1, test_callback2, test_callback3,
                                    test_callback4, test_callback5)
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_callbacks.py:6: ImportError
________________________________________________________________ test_lambda_closure_cleanup ________________________________________________________________

    def test_lambda_closure_cleanup():
>       from pybind11_tests import test_cleanup, payload_cstats
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_callbacks.py:31: ImportError
________________________________________________________________ test_cpp_function_roundtrip ________________________________________________________________

    def test_cpp_function_roundtrip():
        """Test if passing a function pointer from C++ -> Python -> C++ yields the original pointer"""
>       from pybind11_tests import dummy_function, dummy_function2, test_dummy_function, roundtrip
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_callbacks.py:42: ImportError
_________________________________________________________________ test_function_signatures __________________________________________________________________

doc = <conftest.SanitizedString object at 0x7fedcbe959b0>

    def test_function_signatures(doc):
>       from pybind11_tests import test_callback3, test_callback4
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_callbacks.py:60: ImportError
______________________________________________________________________ test_constants _______________________________________________________________________

    def test_constants():
>       from pybind11_tests import some_constant
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_constants_and_functions.py:4: ImportError
_________________________________________________________________ test_function_overloading _________________________________________________________________

    def test_function_overloading():
>       from pybind11_tests import MyEnum, test_function
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_constants_and_functions.py:10: ImportError
________________________________________________________________________ test_bytes _________________________________________________________________________

    def test_bytes():
>       from pybind11_tests import return_bytes, print_bytes
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_constants_and_functions.py:19: ImportError
____________________________________________________________________ test_unscoped_enum _____________________________________________________________________

    def test_unscoped_enum():
>       from pybind11_tests import UnscopedEnum, EOne
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_enum.py:5: ImportError
_____________________________________________________________________ test_scoped_enum ______________________________________________________________________

    def test_scoped_enum():
>       from pybind11_tests import ScopedEnum, test_scoped_enum
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_enum.py:21: ImportError
_________________________________________________________________ test_implicit_conversion __________________________________________________________________

    def test_implicit_conversion():
>       from pybind11_tests import ClassWithUnscopedEnum
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_enum.py:35: ImportError
________________________________________________________________________ test_evals _________________________________________________________________________

capture = <conftest.Capture object at 0x7fedcbea9710>

    def test_evals(capture):
>       from pybind11_tests import (test_eval_statements, test_eval, test_eval_single_statement,
                                    test_eval_file, test_eval_failure, test_eval_file_failure)
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_eval.py:5: ImportError
________________________________________________________________________ test_custom ________________________________________________________________________

msg = <conftest.SanitizedString object at 0x7fedcbebb358>

    def test_custom(msg):
>       from pybind11_tests import (MyException, throws1, throws2, throws3, throws4,
                                    throws_logic_error)
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_exceptions.py:5: ImportError
_____________________________________________________________________ test_inheritance ______________________________________________________________________

msg = <conftest.SanitizedString object at 0x7fedcbec1320>

    def test_inheritance(msg):
>       from pybind11_tests import Pet, Dog, Rabbit, dog_bark, pet_name_species
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_inheritance.py:5: ImportError
_________________________________________________________________ test_automatic_upcasting __________________________________________________________________

    def test_automatic_upcasting():
>       from pybind11_tests import return_class_1, return_class_2, return_none
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_inheritance.py:31: ImportError
_____________________________________________________________________ test_regressions ______________________________________________________________________

    def test_regressions():
>       from pybind11_tests.issues import print_cchar, print_char
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:6: ImportError
____________________________________________________________________ test_dispatch_issue ____________________________________________________________________

msg = <conftest.SanitizedString object at 0x7fedcbeb9940>

    def test_dispatch_issue(msg):
        """#159: virtual function dispatch has problems with similar-named functions"""
>       from pybind11_tests.issues import DispatchIssue, dispatch_issue_go
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:16: ImportError
__________________________________________________________________ test_reference_wrapper ___________________________________________________________________

    def test_reference_wrapper():
        """#171: Can't return reference wrappers (or STL data structures containing them)"""
>       from pybind11_tests.issues import Placeholder, return_vec_of_reference_wrapper
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:37: ImportError
_________________________________________________________________ test_iterator_passthrough _________________________________________________________________

    def test_iterator_passthrough():
        """#181: iterator passthrough did not compile"""
>       from pybind11_tests.issues import iterator_passthrough
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:45: ImportError
____________________________________________________________________ test_shared_ptr_gc _____________________________________________________________________

    def test_shared_ptr_gc():
        """// #187: issue involving std::shared_ptr<> return value policy & garbage collection"""
>       from pybind11_tests.issues import ElementList, ElementA
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:52: ImportError
________________________________________________________________________ test_no_id _________________________________________________________________________

capture = <conftest.Capture object at 0x7fedcb25e6d8>, msg = <conftest.SanitizedString object at 0x7fedcb25e780>

    def test_no_id(capture, msg):
>       from pybind11_tests.issues import get_element, expect_float, expect_int
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:63: ImportError
______________________________________________________________________ test_str_issue _______________________________________________________________________

msg = <conftest.SanitizedString object at 0x7fedcb233e10>

    def test_str_issue(msg):
        """Issue #283: __str__ called on uninitialized instance when constructor arguments invalid"""
>       from pybind11_tests.issues import StrIssue
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:110: ImportError
________________________________________________________________________ test_nested ________________________________________________________________________

    def test_nested():
        """ #328: first member in a class can't be used in operators"""
>       from pybind11_tests.issues import NestA, NestB, NestC, get_NestA, get_NestB, get_NestC
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_issues.py:126: ImportError
_________________________________________________________________ test_keep_alive_argument __________________________________________________________________

capture = <conftest.Capture object at 0x7fedcbe95d30>

    def test_keep_alive_argument(capture):
>       from pybind11_tests import Parent, Child
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_keep_alive.py:5: ImportError
_______________________________________________________________ test_keep_alive_return_value ________________________________________________________________

capture = <conftest.Capture object at 0x7fedcbea9e48>

    def test_keep_alive_return_value(capture):
>       from pybind11_tests import Parent
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_keep_alive.py:39: ImportError
_____________________________________________________________________ test_return_none ______________________________________________________________________

capture = <conftest.Capture object at 0x7fedcb25ebe0>

    def test_return_none(capture):
>       from pybind11_tests import Parent
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_keep_alive.py:73: ImportError
____________________________________________________________________ test_nested_modules ____________________________________________________________________

    def test_nested_modules():
>       import pybind11_tests
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_modules.py:3: ImportError
__________________________________________________________________ test_reference_internal __________________________________________________________________

    def test_reference_internal():
>       from pybind11_tests import ConstructorStats
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_modules.py:13: ImportError
______________________________________________________________________ test_importing _______________________________________________________________________

    def test_importing():
>       from pybind11_tests import OD
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_modules.py:50: ImportError
_____________________________________________________________________ test_string_list ______________________________________________________________________

    def test_string_list():
>       from pybind11_tests import StringList, ClassWithSTLVecProperty, print_opaque_list
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_opaque_types.py:5: ImportError
_______________________________________________________________________ test_pointers _______________________________________________________________________

msg = <conftest.SanitizedString object at 0x7fedcc133908>

    def test_pointers(msg):
>       from pybind11_tests import (return_void_ptr, get_void_ptr_value, ExampleMandA,
                                    print_opaque_list, return_null_str, get_null_str_value,
                                    return_unique_ptr, ConstructorStats)
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_opaque_types.py:27: ImportError
_________________________________________________________________ test_operator_overloading _________________________________________________________________

    def test_operator_overloading():
>       from pybind11_tests import Vector2, Vector, ConstructorStats
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_operator_overloading.py:3: ImportError
________________________________________________________________ test_generalized_iterators _________________________________________________________________

    def test_generalized_iterators():
>       from pybind11_tests import IntPairs
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_sequences_and_iterators.py:14: ImportError
_______________________________________________________________________ test_sequence _______________________________________________________________________

    def test_sequence():
>       from pybind11_tests import Sequence, ConstructorStats
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_sequences_and_iterators.py:26: ImportError
_____________________________________________________________________ test_map_iterator _____________________________________________________________________

    def test_map_iterator():
>       from pybind11_tests import StringMap
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_sequences_and_iterators.py:74: ImportError
______________________________________________________________________ test_vector_int ______________________________________________________________________

    def test_vector_int():
>       from pybind11_tests import VectorInt
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_stl_binders.py:4: ImportError
____________________________________________________________________ test_vector_custom _____________________________________________________________________

    def test_vector_custom():
>       from pybind11_tests import El, VectorEl, VectorVectorEl
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_stl_binders.py:32: ImportError
_____________________________________________________________________ test_vector_bool ______________________________________________________________________

    def test_vector_bool():
>       from pybind11_tests import VectorBool
E       ImportError: generic_type: type "ConstructorStats" is already registered!

test_stl_binders.py:46: ImportError
======================================================= 37 failed, 3 skipped, 9 error in 0.48 seconds =======================================================
make[3]: *** [tests/CMakeFiles/pytest] Error 1
make[2]: *** [tests/CMakeFiles/pytest.dir/all] Error 2
make[1]: *** [tests/CMakeFiles/pytest.dir/rule] Error 2
make: *** [pytest] Error 2
```

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -202,3 +202,27 @@ def pytest_namespace():
         'requires_eigen_and_scipy': skipif(not have_eigen or not scipy,
                                            reason="eigen and/or scipy are not installed"),
     }
+
+
+def _test_import_pybind11():
+    """Early diagnostic for test module initialization errors
+
+    When there is an error during initialization, the first import will report the
+    real error while all subsequent imports will report nonsense. This import test
+    is done early (in the pytest configuration file, before any tests) in order to
+    avoid the noise of having all tests fail with identical error messages.
+
+    Any possible exception is caught here and reported manually *without* the stack
+    trace. This further reduces noise since the trace would only show pytest internals
+    which are not useful for debugging pybind11 module issues.
+    """
+    # noinspection PyBroadException
+    try:
+        import pybind11_tests
+    except Exception as e:
+        print("Failed to import pybind11_tests from pytest:")
+        print("  {}: {}".format(type(e).__name__, e))
+        sys.exit(1)
+
+
+_test_import_pybind11()

--- a/tests/test_eigen.py
+++ b/tests/test_eigen.py
@@ -3,12 +3,11 @@ import pytest
 with pytest.suppress(ImportError):
     import numpy as np
 
-
-ref = np.array([[ 0,  3,  0,  0,  0, 11],
-                [22,  0,  0,  0, 17, 11],
-                [ 7,  5,  0,  1,  0, 11],
-                [ 0,  0,  0,  0,  0, 11],
-                [ 0,  0, 14,  0,  8, 11]])
+    ref = np.array([[ 0,  3,  0,  0,  0, 11],
+                    [22,  0,  0,  0, 17, 11],
+                    [ 7,  5,  0,  1,  0, 11],
+                    [ 0,  0,  0,  0,  0, 11],
+                    [ 0,  0, 14,  0,  8, 11]])
 
 
 def assert_equal_ref(mat):

--- a/tests/test_numpy_dtypes.cpp
+++ b/tests/test_numpy_dtypes.cpp
@@ -268,6 +268,12 @@ py::list test_dtype_methods() {
 }
 
 void init_ex_numpy_dtypes(py::module &m) {
+    try {
+        py::module::import("numpy");
+    } catch (...) {
+        return;
+    }
+
     PYBIND11_NUMPY_DTYPE(SimpleStruct, x, y, z);
     PYBIND11_NUMPY_DTYPE(PackedStruct, x, y, z);
     PYBIND11_NUMPY_DTYPE(NestedStruct, a, b);

--- a/tests/test_numpy_dtypes.py
+++ b/tests/test_numpy_dtypes.py
@@ -2,14 +2,14 @@ import pytest
 with pytest.suppress(ImportError):
     import numpy as np
 
+    simple_dtype = np.dtype({'names': ['x', 'y', 'z'],
+                             'formats': ['?', 'u4', 'f4'],
+                             'offsets': [0, 4, 8]})
+    packed_dtype = np.dtype([('x', '?'), ('y', 'u4'), ('z', 'f4')])
+
 
 def assert_equal(actual, expected_data, expected_dtype):
     np.testing.assert_equal(actual, np.array(expected_data, dtype=expected_dtype))
-
-simple_dtype = np.dtype({'names': ['x', 'y', 'z'],
-                         'formats': ['?', 'u4', 'f4'],
-                         'offsets': [0, 4, 8]})
-packed_dtype = np.dtype([('x', '?'), ('y', 'u4'), ('z', 'f4')])
 
 
 @pytest.requires_numpy

